### DISCUSSION TITLE
Going back without refetching data

### DISCUSSION THREAD

**Discussion by vx17rt** [vercel/next.js#14306] [Help]
Hey there!

I have a small app deployed on Zeit. The app has 2 pages: List of products ("/"), and Product details("/product/[pid]/[oid]")

First page fetches the list of products with a little bit of information about each one (getServerSideProps). Each item is a link to Product details page.
```javascript
<Link
  key={product.id}
  as={`/products/${product.id}/${product.owner_id}`}
  href="/products/[pid]/[oid]"
>
  <a>
    <Product key={product.id} product={product} />
  </a>
</Link>
```

Second page fetches product details and information about product owner (getServerSideProps) and the Back button.
```javascript
<Link href="/" as="/">
  <a>Back</a>
</Link>
```

Questions are: 
~How can I move back from details page to list page without refetching the whole products again?~
~May be something like caching?~
~And how can I preserve scroll position on the first page, when I come back to it from the second?~

**Comment by balazsorban44** [comment]
Setting `shallow` on Link should help. https://nextjs.org/docs/api-reference/next/link

See also: https://nextjs.org/docs/routing/shallow-routing

**Reply by vx17rt** [reply]
Actually it's not, because I'm changing pages during navigation

**Comment by vx17rt** [comment]
Just solved the first two 🙂 by setting Cache Control header, the last question is still the issue, though

**Reply by PixeledCode** [reply]
> 
> 
> Just solved the first two 🙂 by setting Cache Control header, the last question is still the issue, though

I know it's been a long time but did you find any solution for this issue?

**Reply by vx17rt** [reply]
Yep 🙂, you can do something like that in your _app.js

```JavaScript
initRouterListeners();

function initRouterListeners() {
	const routes = [];

	Router.events.on("routeChangeStart", url => {
		pushCurrentRouteInfo();
	});

	Router.events.on("routeChangeComplete", url => {
		window.requestAnimationFrame(() => window.scrollTo(0, 1));
		fixScrollPosition();
	});

	function pushCurrentRouteInfo() {
		routes.push({ pathname: Router.pathname, scrollY: window.scrollY });
	}

	function isBack() {
		return routes.length >= 2 && Router.pathname === routes[routes.length - 2].pathname;
	}

	function fixScrollPosition() {
		let scrollY = 0;

		if (isBack()) {
			routes.pop();
			const targetRoute = routes.pop();
			scrollY = targetRoute.scrollY;
		}

		window.requestAnimationFrame(() => window.scrollTo(0, scrollY));
	}
}

**Reply by hpeinar** [reply]
@tourervit would it be possible to elaborate on that `Cache-control` header solution?

Did you actually solve the re-rendering or you are just caching the API response for the homepage so the data fetch is instant?

Because we have same problem but our front page is using infinite scroll and can have hundres of products loaded from multiple API calls, even if the API calls are cached then the re-render will take considerable amount of time and is noticeable by the user.

So I'm looking ways to go "Back" without doing any of the rendering again.

**Reply by vx17rt** [reply]
Sorry, can't run that project... I no longer have an access to it, and API is dead, but I checked a messenger where I shared code with my colleague, here what I can share:

```javascript
export const getServerSideProps: GetServerSideProps = async ctx => {
	ctx.res.setHeader("Cache-Control", "s-maxage=1, stale-while-revalidate");

	// here was fetching logic

	return {
		props: {
			products,
		},
	};
};
```
That was Index page.
So I guess, there was just caching and I have to mention that we were using Vercel as our hosting platform.
